Rich text markdown editor

GAsplund opened the pull request:
# Key Features
- Add rich text markdown editor through Milkdown
- Includes a source editing and preview mode

# Known Issues
- You cannot use the mouse cursor to select text in an image upload pane. Keyboard shortcuts only.
- Some buttons have subpar contrast

# Screenshots
<img width="1060" height="294" alt="Rich text editor screen" src="https://github.com/user-attachments/assets/6711647d-4469-4369-9d59-354eb1af5e3d" />

GAsplund committed:
Use MDXEditor

GAsplund committed:
Switch to Milkdown editor

GAsplund committed:
Merge tag 'v1.2.8' into feature/rich-editor

GAsplund committed:
Allow file uploads

GAsplund committed:
Improve styling and behavior

GAsplund committed:
Add locale

GAsplund committed:
Use local files for image previews

GAsplund committed:
Fix loader being triggered when clicking link

GAsplund committed:
Improve previews

GAsplund committed:
Fix ESLint, improve styling

GAsplund committed:
Decrease code repetition

GAsplund committed:
Improve link copying behavior

GAsplund committed:
Add missing dependency

GAsplund committed:
Merge branch 'main' into feature/rich-editor

Sus-H commented:
Looks beautiful, however I am having some issues with the link button. It worked the first time i tried it and then for some reason nothing happened when I clicked it. Attaching image with console in case it helps.

Also, when editing a long text, it would be nice if the toolbar was always visible, because i need to scroll up in case I need to use something from it.

<img width="1709" height="909" alt="image" src="https://github.com/user-attachments/assets/f903ce72-807a-4a3f-aded-2820584cff6c" />

GAsplund commented:
> Looks beautiful, however I am having some issues with the link button. It worked the first time i tried it and then for some reason nothing happened when I clicked it. Attaching image with console in case it helps.
> 
> Also, when editing a long text, it would be nice if the toolbar was always visible, because i need to scroll up in case I need to use something from it.

Are you talking about the fact that clicking the link button without any text selected does nothing? I think I could add some default behavior for that to make it always behave as expected.

I'll look into fixing the toolbar. Thanks for the feedback!

Sus-H commented:
Ohhh yeah sorry, I forgot it worked that way. hmm...

GAsplund committed:
Make sure that editor is loaded

GAsplund committed:
Add placeholder link if no selection

Sus-H commented:
Something his happening when I try to make a list by typing dash and then space as well, the dash disappears sometimes


https://github.com/user-attachments/assets/f1097de1-4da2-4e21-812e-331e91c1a6bd

Sus-H commented:
Another feedback, WYSIWYG for the different heading sizes would be nice!

Sorry for spam

<img width="553" height="349" alt="image" src="https://github.com/user-attachments/assets/1d7cac15-1ef5-479b-ae29-a522a6e5487c" />

Sus-H commented:
Another issue: I just clicked the richtext button (when it was already selected) and all my content disappeared.

<img width="1055" height="698" alt="image" src="https://github.com/user-attachments/assets/fc72cfbd-e1d5-4aef-ad0e-54353f01f7f4" />

It seems like clicking it while its already selected removes the content, it was possible to undo it unless you selected another view.

Sus-H commented:
Another thing: I'm suspecting its the empty file upload and something else, but yeah, but they show as "1.00" in the preview mode which can be a bit confusing. edit: Oh I can see its the link text when i look in markdown view.
<img width="930" height="927" alt="image" src="https://github.com/user-attachments/assets/655ccae3-b395-4a1c-8905-368351b3e42b" />

<img width="921" height="486" alt="image" src="https://github.com/user-attachments/assets/e958a5dd-1fed-47c4-94d3-3e8c2efb64e4" />

GAsplund committed:
Merge branch 'feature/rich-editor' of https://github.com/cthit/chalmers.it into feature/rich-editor

GAsplund committed:
Sticky editor toolbar

GAsplund committed:
Use same styling as preview

GAsplund committed:
Improve viewMode logic

GAsplund commented:
Responding to all your points here:
- I've changed the link button to create a text field with "text", on which it adds a new link as if you selected it
- Bullet points seem like a CSS issue. I've tried to make the CSS consistent between the editor and preview and it is almost identical to it now from my testing
- It seems like you'd need to make a custom drop-down renderer in order to be able to do custom CSS on options. The options themselves are apparently rendered natively by the OS itself
- It seems like content is only removed when clicking it after not having previously selected anything. I've changed the logic to do nothing if you try to switch from one mode to itself which seemed to fix it.
- I'm unable to reproduce images showing in WYSIWYG but not preview. Any local images show fine for me. However, I've found the preview in WYSIWYG to be a bit spotty for local images at times. Will look further into it!

GAsplund committed:
Merge remote-tracking branch 'origin/main' into feature/rich-editor

GAsplund committed:
Merge remote-tracking branch 'origin/main' into feature/rich-editor

Sus-H commented:
- There is a small gap above the stickied toolbar. Also maybe set the div with tools beside the title of the post? Title and post are grouped in my mind so it would make sense they would be visually as well.
<img width="967" height="149" alt="image" src="https://github.com/user-attachments/assets/b5edcec1-f5c8-42c8-a5ac-6b980fd69358" />

- The toolbar should be mobile friendly / responsive

- We talked about the code block not being visually different from paragraph text - maybe give it a different background. It was confusing trying to add a link into it.
<img width="530" height="300" alt="image" src="https://github.com/user-attachments/assets/a2133dd8-6a4e-41d3-b72d-43b14f4625cd" />


- image caption: would be nice if it disappeared(toggled off) when you backspaced in an empty caption box
<img width="382" height="352" alt="image" src="https://github.com/user-attachments/assets/c7fff9e2-5e5e-44a8-95c3-8ee9db52f92a" />



- im not sure how i did this but the caption button and link input started floating when i scrolled
<img width="834" height="720" alt="image" src="https://github.com/user-attachments/assets/edd47f85-f472-4cf8-abce-d086d8d7378d" />

here the link input disappeared but the caption button is still floating even when I tried to reset it by clicking it. Only resetted when i removed the image and added it again.
<img width="463" height="426" alt="image" src="https://github.com/user-attachments/assets/6042891b-4cb6-411d-9415-f35543dcebbb" />

<img width="487" height="439" alt="image" src="https://github.com/user-attachments/assets/2f76a65f-b55e-4881-acc1-3e7b642dc422" />

- It would be nice if there was a cursor (the vertical blinking line) to the right of an image instead of marking/highlighting the image when you try to place the cursor to the right of it

GAsplund committed:
Merge remote-tracking branch 'origin/main' into feature/rich-editor

GAsplund committed:
Custom serializer for images

GAsplund committed:
Minor style changes

GAsplund commented:
I've made the following changes:
- Dropdown background is !important and the toolbar now sticks to the header
<img width="1089" height="213" alt="image" src="https://github.com/user-attachments/assets/c4f1b1e2-7cb2-4ce5-b40d-bfee95226389" />

- Images now serialize and deserialize as expected. This should solve most 1.00 issues and make the generated Markdown more intuitive

Things that I have not addressed are:
- Mobile/thin screen support
- Stating the current text type (paragraph, link, title, etc.)

I think these changes are a bit too big/annoying for this PR as of now and should probably be added later.

GAsplund committed:
Prevent image blocks from creeping up

GAsplund committed:
Auto-fill group ID when editing

Disabled when editing a news post. Does not affect when saving anyway.

Sus-H reviewed (approved):
It was discussed to add this new editor as it is now and then add further improvements since it is already better than the current one.

GAsplund committed:
Merge branch 'main' into feature/rich-editor
